File: JavaScript/src/items/Item.js
// Represents a single unit of inventory
export default class Item {
  constructor(gildedRose) {
    this.gildedRose = gildedRose;
  }

  tick() {
    throw new Error('Must be implemented by subclass');
  }

  ensureBoundedQuality() {
    if (this.gildedRose.quality < 0) {
      this.gildedRose.quality = 0;
    }

    if (this.gildedRose.quality > 50) {
      this.gildedRose.quality = 50;
    }
  }
}

File: Ruby/lib/items/item.rb
# frozen_string_literal: true

module Items
  # Represents a single unit of inventory
  class Item
    def initialize(gilded_rose)
      @gilded_rose = gilded_rose
    end

    def tick
      raise('Must be implemented by subclass')
    end

    private def ensure_bounded_quality
      @gilded_rose.quality = 0 if @gilded_rose.quality < 0
      @gilded_rose.quality = 50 if @gilded_rose.quality > 50
    end
  end
end
